contracts/test/LON/Permit.t.sol

```diff
@@ -0,0 +1,115 @@
+    }
+
+    function testCannotPermitByZeroAddress() public {
+        Permit memory permit = DEFAULT_PERMIT;
+        permit.owner = address(0);
+        (uint8 v, bytes32 r, bytes32 s) = _signPermit(userPrivateKey, permit);
+
+        vm.expectRevert("zero address");
+        lon.permit(permit.owner, permit.spender, permit.value, permit.deadline, v, r, s);
+    }
+
+    function testCannotPermitWhenExpired() public {
+        Permit memory permit = DEFAULT_PERMIT;
+        permit.deadline = block.timestamp - 1;
+        (uint8 v, bytes32 r, bytes32 s) = _signPermit(userPrivateKey, permit);
+
+        vm.expectRevert("permit is expired");
+        lon.permit(permit.owner, permit.spender, permit.value, permit.deadline, v, r, s);
+    }
+
+    function testCannotPermitWithInvalidSignature() public {
+        Permit memory permit = DEFAULT_PERMIT;
+        (uint8 v, bytes32 r, bytes32 s) = _signPermit(otherPrivateKey, permit);
+
+        vm.expectRevert("invalid signature");
+        lon.permit(permit.owner, permit.spender, permit.value, permit.deadline, v, r, s);
+    }
+
+    function testCannotPermitIfSignAgain() public {
+        Permit memory permit = DEFAULT_PERMIT;
+        (uint8 v, bytes32 r, bytes32 s) = _signPermit(userPrivateKey, permit);
+
+        lon.permit(permit.owner, permit.spender, permit.value, permit.deadline, v, r, s);
+
+        vm.expectRevert("invalid signature");
+        lon.permit(permit.owner, permit.spender, permit.value, permit.deadline, v, r, s);
+    }
+
+    function testPermit() public {
+        Permit memory permit = DEFAULT_PERMIT;
+        (uint8 v, bytes32 r, bytes32 s) = _signPermit(userPrivateKey, permit);
+
+        uint256 nonceBefore = lon.nonces(user);
+        uint256 allowanceBefore = lon.allowance(permit.owner, permit.spender);
+        lon.permit(permit.owner, permit.spender, permit.value, permit.deadline, v, r, s);
+        uint256 nonceAfter = lon.nonces(user);
+        uint256 allowanceAfter = lon.allowance(permit.owner, permit.spender);
+
+        assertEq(nonceAfter, nonceBefore + 1);
+        assertEq(allowanceAfter, allowanceBefore + permit.value);
+    }
+
+    /*********************************
+     *          Test Helpers         *
+     *********************************/
+
+    function _signPermit(uint256 privateKey, Permit memory permit)
+        internal
+        returns (
+            uint8,
+            bytes32,
+            bytes32
+        )
+    {
+        bytes32 permitHash = _getPermitHash(permit);
+        bytes32 EIP712SignDigest = _getEIP712Hash(permitHash);
+
+        (uint8 v, bytes32 r, bytes32 s) = vm.sign(privateKey, EIP712SignDigest);
+        return (v, r, s);
+    }
+
+    function _getEIP712Hash(bytes32 structHash) internal view returns (bytes32) {
+        string memory EIP191_HEADER = "\x19\x01";
+        bytes32 DOMAIN_SEPARATOR = lon.DOMAIN_SEPARATOR();
+        return keccak256(abi.encodePacked(EIP191_HEADER, DOMAIN_SEPARATOR, structHash));
+    }
```

Comment:
I feel this is a general util not only for permit. Maybe this is more suitable to place in setup.

Reply:
Yes! I also feel that way. I've Moved the _getEIP712Hash general util function from the Permit.t.sol test file to Setup.t.sol test contract.